### DISCUSSION TITLE
[atari5200] Atari 5200 color definitions that match Atari 8-bit definitions?

### DISCUSSION THREAD

**Discussion by Fabrizio-Caruso** [cc65/cc65#1364] [Q&A] [answered]
I am trying to have Atari 5200 colors match Atari 8-bit colors but I get different colors on the two very similar targets.

Which color definitions should I use?
TGI / GTIA definitions?
Only the COLOR_... definitions?
If I want to choose the very specific hue/luma?

On the Atari 8-bit something like:
        
        _setcolor_low(0, GTIA_COLOR_RED);
        _setcolor_low(1, TGI_COLOR_WHITE);
        _setcolor_low(2, GTIA_COLOR_GREEN); 
        _setcolor_low(3, GTIA_COLOR_BROWN);
        _setcolor_low(4, TGI_COLOR_BLACK);
        
seems to work.

How do I get the very same colors on the Atari 5200?

**Comment by oliverschmidt** [comment]
@groessler: A question maybe you want to help answering...

**Comment by groessler** [comment, marked as answer]
On 12/18/20 11:55 PM, Fabrizio Caruso wrote:
>
> How do I get the very same colors on the Atari 5200?
>

Untested on hw or emulator:

#define SETCOLOR_LOW(reg, val) *((unsigned char *)12 + (reg)) = (val)


e.g.

// test 5200

#define SETCOLOR_LOW(reg, val) *((unsigned char *)12 + (reg)) = (val)

#include <atari5200.h>

int main(void)
{
         SETCOLOR_LOW(0, GTIA_COLOR_RED);
         SETCOLOR_LOW(1, GTIA_COLOR_WHITE);
         SETCOLOR_LOW(2, GTIA_COLOR_GREEN);
         SETCOLOR_LOW(3, GTIA_COLOR_BROWN);
         SETCOLOR_LOW(4, GTIA_COLOR_BLACK);

         return 0;
}

regards,
chris

**Comment by Fabrizio-Caruso** [comment]
@groessler Colors are different:
A5200:
![image](https://user-images.githubusercontent.com/29057615/103135123-6ce6f680-46b6-11eb-9618-0112cae77943.png)


A8:
![image](https://user-images.githubusercontent.com/29057615/103135168-ba636380-46b6-11eb-81cb-3b008d710fb2.png)

**Comment by groessler** [comment]
On 12/25/20 1:39 PM, Fabrizio Caruso wrote:

> @groessler <https://github.com/groessler> Colors are different:
>

I don't see any difference. But I'm color-blind.


Which emulator do you use?

Use the same video system (NTSC) for both versions. Make sure that 
artifacting modes and emulator palettes are the same.


regards,
chris

**Comment by Fabrizio-Caruso** [comment]
@groessler, if you are color blind you may not be able to see the difference between green walls (above) and brown walls (below) but you should see dark apples (above) and clearer apples (below). Also the snake color color is not the same green (darker above and clearer below).

I am using the most accurate emulator: Altirra (3.20).

**Comment by groessler** [comment]
Looking harder I'm seeing it. But you also had a different pic for A8 in your original post which I answered from mail.

Did you set every video related setting to the same in the emulator?

**Comment by Fabrizio-Caruso** [comment]
@groessler problem found! My mistake. A5200 is only NTSC.... whereas my A8 config was PAL... 
Sorry guys for this.
I now need to figure out how to have clearer color on NTSC output.

**Comment by groessler** [comment]
The 16 predefined COLOR values are just 16 out of possible 128 colors, selected to be similar to the C64.

`setcolor_low` can use any color. See in `gtia.h`:

```
/* Make a GTIA color value */
#define _gtia_mkcolor(hue,lum) (((hue) << 4) | ((lum) << 1))

/* Luminance values go from 0 (black) to 7 (white) */

/* Hue values */
/* (These can vary depending on TV standard (NTSC vs PAL),
** tint potentiometer settings, TV tint settings, emulator palette, etc.)
*/
#define HUE_GREY        0
#define HUE_GOLD        1
#define HUE_GOLDORANGE  2
#define HUE_REDORANGE   3
#define HUE_ORANGE      4
#define HUE_MAGENTA     5
#define HUE_PURPLE      6
#define HUE_BLUE        7
#define HUE_BLUE2       8
#define HUE_CYAN        9
#define HUE_BLUEGREEN   10
#define HUE_BLUEGREEN2  11
#define HUE_GREEN       12
#define HUE_YELLOWGREEN 13
#define HUE_YELLOW      14
#define HUE_YELLOWRED   15
```

**Comment by Fabrizio-Caruso** [comment]
@groessler thanks a lot!
Ok, I think I need to check at run time if I am on a PAL or NTSC A8. A5200 is always NTSC, so it is easier.
How do I use `__gtia_read`  to access the `pal` flag?

**Comment by groessler** [comment]
You can use CC65's `get_tv()` function which is also available on other targets.
